### PR TITLE
Refactor MultiDeviceExecutor to allow separate initialization

### PR DESCRIPTION
**Pull request by jjsjann123**

### CODE REVIEW

File: csrc/multidevice/executor.h [hunk 1]
```diff
@@ -123,7 +136,7 @@ class MultiDeviceExecutor {
   std::unordered_map<Val*, c10::IValue> val_to_IValue_;
 
   // holds the Communicator to be used for execution
-  Communicator& comm_;
+  Communicator* comm_ = nullptr;
```
Comment (jjsjann123) [inline csrc/multidevice/executor.h:139]: I'm wondering if this should have been a `std::shared_ptr<Communicator>` instead.

**Comment by jjsjann123** [comment]
!build --dist

**Comment by cowanmeg** [comment]
> !build --dist

Let me merge the test fixes in. Right now a UCC test will hang everything xD

**Comment by jjsjann123** [comment]
> > !build --dist
> 
> Let me merge the test fixes in. Right now a UCC test will hang everything xD

Thanks for the heads up. I shouldn't have started the CI then. :fearful:

**Comment by jjsjann123** [comment]
!build --dist

**Comment by jjsjann123** [comment]
Now I think I have your test fixes :sweat_smile:

**Review by samnordmann — COMMENTED** [review]
Can you give a bit of context about why this patch is needed?

As far as I understand, the patch is made to remove the bool flag `multidevice_flag` in the python API in the [other pr](https://github.com/NVIDIA/Fuser/pull/1935#discussion_r1554151585), right ?
However the current patch is imo a greater increase in complexity, so I am not sure if it is relevant... Indeed, we still pay the price of a boolean flag `is_initialized_` anyway, plus the fact that the new`init` method must be called before any other call (so naturally belongs to the constructor and should not be a standalone public method imo).

 On the other hand, having an explicit flag to turn on/off multidevice in the python API seems reasonable and handy for now.

**Comment by cowanmeg** [comment]
> Can you give a bit of context about why this patch is needed?
> 
> As far as I understand, the patch is made to remove the bool flag `multidevice_flag` in the python API in the [other pr](https://github.com/NVIDIA/Fuser/pull/1935#discussion_r1554151585), right ? However the current patch is imo a greater increase in complexity, so I am not sure if it is relevant... Indeed, we still pay the price of a boolean flag `is_initialized_` anyway, plus the fact that the new`init` method must be called before any other call (so naturally belongs to the constructor and should not be a standalone public method imo).
> 
> On the other hand, having an explicit flag to turn on/off multidevice in the python API seems to me reasonable and handy for now.

It seems like this deferred initialization might be useful for @wujingyue multidevice executor refactor, since communicator could be optional for single device test?

File: csrc/multidevice/executor.cpp [hunk 2]
```diff
@@ -234,19 +243,19 @@ std::vector<at::Tensor> MultiDeviceExecutor::runWithInput(
 }
 
 std::string MultiDeviceExecutor::validate() const {
-  if (!comm_.is_available()) {
+  if (!comm_->is_available()) {
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply csrc/multidevice/executor.cpp:246]: Got it. Will handle that across the places..

**Comment by jjsjann123** [comment]
> > Can you give a bit of context about why this patch is needed?
> > As far as I understand, the patch is made to remove the bool flag `multidevice_flag` in the python API in the [other pr](https://github.com/NVIDIA/Fuser/pull/1935#discussion_r1554151585), right ? However the current patch is imo a greater increase in complexity, so I am not sure if it is relevant... Indeed, we still pay the price of a boolean flag `is_initialized_` anyway, plus the fact that the new`init` method must be called before any other call (so naturally belongs to the constructor and should not be a standalone public method imo).
> > On the other hand, having an explicit flag to turn on/off multidevice in the python API seems to me reasonable and handy for now.
> 
> It seems like this deferred initialization might be useful for @wujingyue multidevice executor refactor, since communicator could be optional for single device test?

Agreeing with @cowanmeg . I think having a deferred initialization would just make it easier to squeeze it in integration at other places.
Otherwise, any place that needs to handle an optional MultiDeviceExecutor is going to need *two* flags for monitoring its initialization, which feels like a lack of feature in the class itself.

**Comment by jjsjann123** [comment]
@wujingyue agrees with @samnordmann on the added complexity of delayed init. He mentioned that he might not need this for integration, so there's no more strong push to pursue this PR.

I'm closing this one and will refactor my old PR to simplify the executor choice.

**Pull request closed without merging**